cranelift/egraph/src/egraph.rs

```diff
@@ -0,0 +1,516 @@
+//! ægraph (aegraph, or acyclic e-graph) implementation.
+//!
+//! An aegraph is a form of e-graph. We will first describe the
+//! e-graph, then the aegraph as a slightly less powerful but highly
+//! optimized variant of it.
+//!
+//! ## The e-graph
+//!
+//! An e-graph, or equivalence graph, is a kind of node-based
+//! intermediate representation (IR) data structure that consists of
+//! *eclasses* and *enodes*. An eclass contains one or more enodes;
+//! semantically an eclass is like a value, and an enode is one way to
+//! compute that value. If several enodes are in one eclass, the data
+//! structure is asserting that any of these enodes, if evaluated,
+//! would produce the value.
+//!
+//! An e-graph also contains a deduplicating hash-map of nodes, so if
+//! the user creates the same e-node more than once, they get the same
+//! e-class ID.
+//!
+//! In the usual use-case, an e-graph is used to build a sea-of-nodes
+//! IR for a function body or other expression-based code, and then
+//! *rewrite rules* are applied to the e-graph. Each rewrite
+//! potentially introduces a new e-node that is equivalent to an
+//! existing e-node, and then unions the two e-nodes' classes
+//! together.
+//!
+//! In the trivial case this results in an e-class containing a series
+//! of e-nodes that are newly added -- all known forms of an
+//! expression -- but Note how if a rewrite rule rewrites into an
+//! existing e-node (discovered via deduplication), rewriting can
+//! result in unioning of two e-classes that have existed for some
+//! time.
+//!
+//! An e-graph's enodes refer to *classes* for their arguments, rather
+//! than other nodes directly. This is key to the ability of an
+//! e-graph to canonicalize: when two e-classes that are already used
+//! as arguments by other e-nodes are unioned, all e-nodes that refer
+//! to those e-classes are themselves re-canonicalized. This can
+//! result in "cascading" unioning of eclasses, in a process that
+//! discovers the transitive implications of all individual
+//! equalities. This process is known as "equality saturation".
+//!
+//! ## The acyclic e-graph (aegraph)
```

Comment:
I added a section to the doc-comment to address this; but in brief, yep, it's basically the example that you give.

It's pretty subtle though, as this only matters if the rewrite from class3 to class1 exists *and if class3 is cheaper*. In other words, if a rewrite takes a cheaper expression and rewrites it into a more expensive expression. Then it may happen to intern with an existing copy of that more expensive expression and union the cheaper version into it. That seems unlikely for the sorts of optimizations we plan to write. In the fullness of time this would still be a good question to investigate further, though!